influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +115,208 @@ struct CacheValue {
+        data.len()
+            + location.as_ref().len()
+            + e_tag.as_ref().map(|s| s.capacity()).unwrap_or_default()
+            + version.as_ref().map(|s| s.capacity()).unwrap_or_default()
+    }
+}
+
+#[derive(Debug)]
+struct CacheEntry {
+    state: CacheEntryState,
+    /// A counter for tracking how many times this entry has been hit.
+    ///
+    /// When first created in the `Fetching` state, this will be set to -1, which will prevent
+    /// the fetching entry from being evicted by a prune operation
+    hits: AtomicI8,
+}
+
+impl CacheEntry {
+    /// Increment the used counter for this entry unless it is already at the maximum value
+    fn increment_hits(&self) {
+        let _ = self
+            .hits
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |x| {
+                if x == i8::MAX {
+                    None
+                } else {
+                    Some(x + 1)
+                }
+            });
+    }
+
+    /// Get the approximate memory footprint of this entry in bytes
+    fn size(&self) -> usize {
+        self.state.size() + std::mem::size_of::<AtomicU8>()
     }
 }
 
 /// The state of a cache entry
 #[derive(Debug)]
-enum CacheEntry {
+enum CacheEntryState {
     /// The cache entry is being fetched from object store
-    Fetching,
+    ///
+    /// This holds a [`watch::Sender`] that is used to notify requests made to get this entry
+    /// while it is being fetched by the cache oracle.
+    Fetching(watch::Sender<Option<Arc<CacheValue>>>),
     /// The cache entry was successfully fetched and is stored in the cache as a [`CacheValue`]
     Success(Arc<CacheValue>),
-    /// The request to the object store failed
-    Failed,
-    /// The cache entry was deleted
-    Deleted,
-    /// The object is too large for the cache
-    TooLarge,
 }
 
-impl CacheEntry {
-    /// Get the size of thje cache entry in bytes
+impl CacheEntryState {
+    /// Get the approximate size of the cache entry in bytes
     fn size(&self) -> usize {
         match self {
-            CacheEntry::Fetching => 0,
-            CacheEntry::Success(v) => v.size(),
-            CacheEntry::Failed => 0,
-            CacheEntry::Deleted => 0,
-            CacheEntry::TooLarge => 0,
+            CacheEntryState::Fetching(tx) => std::mem::size_of_val(tx),
+            CacheEntryState::Success(v) => v.size(),
         }
     }
+}
+
+/// A cache for storing objects from object storage by their [`Path`]
+#[derive(Debug)]
+struct Cache {
+    /// The maximum amount of memory this cache should occupy
+    capacity: usize,
+    /// The current amount of memory being used by the cache
+    used: AtomicUsize,
+    /// The maximum amount of memory to free during a prune operation
+    max_free_amount: usize,
+    /// The map storing cache entries
+    ///
+    /// This uses [`IndexMap`] to preserve insertion order, such that, when iterating over the map
+    /// to prune entries, older entries are removed before newer entries
```

Comment:
Clarify that this is in the event of a tie - because the iteration will process entries in the order that they were originally inserted, i.e., fetched for the cache.